Updated the learn2learn "image_classification_imagenette_mini" example

uakarsh opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
 Please also include relevant motivation and context.
 List any dependencies that are required for this change.
-->

There were some problems in the downloading of the dataset as well as defining the `ImageClassificationInputTransform`, for the `ImageClassificationData`, so I have remade the tutorial for easy integration of `learn2learn` with `flash`
Fixes #1376

## Before submitting
- [ ] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the **[contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md)**, Pull Request section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? [not needed for typos/docs]
- [ ] Did you verify **new and existing tests pass** locally with your changes?
- [ ] If you made a notable change (that affects users), did you **update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)**?

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## PR review
 - [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Definitely
Make sure you had fun coding 🙃

krshrimali commented:
Hi, @uakarsh - Thank you for the PR, I appreciate you taking a look at it. Apologies for the delay in review.

Would you mind converting this into a script instead? We usually prefer having python scripts in `flash_examples`, and any notebooks currently go in [`flash_notebooks` folder](https://github.com/Lightning-AI/lightning-flash/tree/master/flash_notebooks). This will also help us review the script, and add suggestions.

A few points:

1. I guess you changed the batch sizes to `1` for testing your PR, so can you revert those parameters to the original? (unless there was another reason that I'm missing)
2. Any reason you changed the tthe ransforms applied for training stage by adding `train_per_sample_transform`? Just curious, I noticed that we didn't have those before (please correct me if I'm wrong).
3. I see that you have gotten rid of `ddp_shared` accelerator which I think makes sense for examples. cc: @ethanwharris - I think it's a good idea to not have multi-gpu enabled by default in examples?

Also, just FYI:

We made a few changes to the `transform` API, if I'm not wrong, it now takes `transform= ImageClassificationInputTransform` in your `ImageClassificationData.from_tensors` call (instead of `train_transform` and `val_transform` kwargs). Whether you need to make this change in this PR or not, depends on which release this PR land for. I think @ethanwharris can comment more on this. :)

uakarsh commented:
Sure, I can convert it into a script

For your questions:
1. The reason for using batch size 1 is, that I was using Amazon Sagemaker Studio (and, taking a batch size of more than 1, would lead to CUDA OUT OF MEMORY ERROR, so I had no other option but to use a batch size of 1 for making the code run

2. It was taken from [here](https://lightning-flash.readthedocs.io/en/stable/reference/image_classification.html#custom-transformations)

3. I removed it because I think it is not useful for people having only a single GPU (and I rely only on Colab/Kaggle/Sagemaker Studio, so for that purpose I removed the DDP strategy) for coding/ removing the bugs

ethanwharris commented:
Hey @uakarsh instead of creating a script from scratch, could you just apply your changes to the script here: https://github.com/Lightning-AI/lightning-flash/blob/master/flash_examples/integrations/learn2learn/image_classification_imagenette_mini.py

Happy for you to drop the `ddp_shared`, looks like a typo anyway as I guess it was meant to be `ddp_sharded`, not too sure.

Hope that helps :smiley:

krshrimali commented:
Hey, @uakarsh - Any updates on this? :) Let us know if you need any help.

uakarsh commented:
Would surely do it today. Got a bit engaged in my own work and about using PyTorch Lightning with Speech Recongition.

uakarsh commented:
Update: Have updated the script (along with instructions for downloading the dataset) and removed the notebook.

krshrimali commented on flash_examples/integrations/learn2learn/image_classification_imagenette_mini.py:
```diff
@@ -14,71 +14,139 @@
 
 # adapted from https://github.com/learnables/learn2learn/blob/master/examples/vision/protonet_miniimagenet.py#L154
 
+"""
+Requirements:
+
+pip install learn2learn
+pip install kornia
+pip install lightning-flash
+pip install 'lightning-flash[image]'
+"""
+
```
Thanks for adding this here. I understand the motivation for adding this, but let's remove it as we expect users to install flash, kornia and learn2learn if they are using it. If they are running this script from the flash repository, they can do `pip install -r requirements/datatype_image_extras.txt".

```suggestion
```

krshrimali commented on flash_examples/integrations/learn2learn/image_classification_imagenette_mini.py:
```diff
@@ -14,71 +14,139 @@
+        return T.Compose(
+            [
+                ApplyToKeys(
+                    "input",
```
minor nit for consistency:

```suggestion
                    DataKeys.INPUT,
```

krshrimali reviewed (approved):
Nice attention to detail, @uakarsh - thanks for fixing the example script. We appreciate it!

I'm only wondering if this will fail the CI since we expect the user to download the dataset. I'll keep an eye on the CI, and add a custom download function until https://github.com/learnables/learn2learn/issues/310 is fixed.

Thanks! 🎉

uakarsh commented:
Sure, I have added the entry for the updation in [CHANGELOG.md](https://github.com/Lightning-AI/lightning-flash/blob/master/CHANGELOG.md) The change is [here](https://github.com/Lightning-AI/lightning-flash/compare/master...uakarsh:lightning-flash:patch-1)

krshrimali commented on flash_examples/integrations/learn2learn/image_classification_imagenette_mini.py:
```diff
@@ -14,71 +14,127 @@
+"""
+
 import warnings
+from dataclasses import dataclass
+from typing import Callable, Tuple, Union
 
 import kornia.augmentation as Ka
 import kornia.geometry as Kg
 import learn2learn as l2l
+import numpy as np
 import torch
 import torchvision
+import torchvision.transforms as T
+from PIL import Image
 from torch import nn
```
Just some unused imports:

```suggestion
from dataclasses import dataclass
from typing import Tuple, Union

import kornia.augmentation as Ka
import kornia.geometry as Kg
import learn2learn as l2l
import torch
import torchvision.transforms as T
```

krshrimali commented:
Failing tests are irrelevant to this PR. cc: @ethanwharris - can you please merge this whenever you are back?

@uakarsh - Apologies for the delay in getting this merged. The CI was failing for a couple of weeks, and we had to prioritize fixing it! As always, we appreciate your contribution. 🎉

uakarsh commented:
Hi @krshrimali, thanks for merging this. This is my first contribution to any issue on Github, so thanks for giving me this chance and looking forward to helping/resolving more issues.

krshrimali commented:
> Hi @krshrimali, thanks for merging this. This is my first contribution to any issue on Github, so thanks for giving me this chance and looking forward to helping/resolving more issues.

It's always a memory, the first contribution! I still remember the exact time when I got my first PR merged, wish you the best! If you ever find an issue you find exciting to work upon, feel free to ping me, and I'll be happy to help wherever I can 🎉 ⚡
